packages/block-editor/src/components/block-draggable/index.js

```diff
@@ -61,6 +86,64 @@ const BlockDraggable = ( {
 		};
 	}, [] );
 
+	// Find the root of the editor iframe.
+	const blockRef = useBlockRef( clientIds[ 0 ] );
+	const editorRoot = blockRef.current?.closest( 'body' );
+
+	// Add a dragover event listener to the editor root to track the blocks being dragged over.
+	// The listener has to be inside the editor iframe otherwise the target isn't accessible.
+	// Check if the dragged blocks are allowed inside the target. If not, grey out the draggable.
+	useEffect( () => {
```

Comment:
Is this `useEffect` and its `onDragOver` still needed? Since the `useBlockDropZone` is hiding the insertion point, could we replace the whole useEffect with something like:

```
	useEffect( () => {
		if ( ! visibleInserter ) {
			window?.document?.body?.classList?.add(
				'block-draggable-invalid-drag-token'
			);
		} else {
			window?.document?.body?.classList?.remove(
				'block-draggable-invalid-drag-token'
			);
		}
	}, [ visibleInserter ] );
```

Then the logic of when to show and hide the insertion point could live in one place, so that the `BlockDraggable` doesn't have to also perform checks with its own drag event handler.

Or is there another reason why this component needs to perform its own separate checks?

Reply:
Yeah, so the crucial bit here is finding the `targetClientId`, which tells us the container that the draggable is currently hovering over. To find that we need a `dragover` event listener attached to the inside of the editor iframe (if it's on the outside, it won't be able to access the elements inside the iframe). We need the `targetClientId` to check if the drop target is valid, which is necessary because if the chip is hovering over a valid container but there's no drop indicator currently showing (if we're moving between drop areas for instance), we don't want it to grey out.

Unfortunately there's not much that can be done to restrict how many times this runs without slowing down the responsiveness of the drag chip to what it's hovering over. The problem here is that, unlike with the drop indicators that exist statically between all blocks so we know where they're placed from the start, we need to find out what the draggable is being dragged over, and afaik the only way to do that is through the event target of a `dragover` listener.

Reply:
> We need the targetClientId to check if the drop target is valid, which is necessary because if the chip is hovering over a valid container but there's no drop indicator currently showing (if we're moving between drop areas for instance), we don't want it to grey out.

Ah, gotcha! Thanks for the explanation, that makes sense now 👍

> Unfortunately there's not much that can be done to restrict how many times this runs without slowing down the responsiveness of the drag chip to what it's hovering over.

Yeah, it can be really tricky to get the right balance when slowing things down so that it's slow enough for performance reasons, but fast enough to feel responsive to a user! 😅

> we need to find out what the draggable is being dragged over, and afaik the only way to do that is through the event target of a dragover listener.

That sounds about right to me, too. If it winds up being a bit of a rabbithole trying to get it working smoothly, would it be worth splitting out the `useBlockDropZone` change and landing that first? Totally cool if you want to keep it all together while figuring everything out of course!